sample_size/sample_size_calculator.py

```diff
@@ -0,0 +1,106 @@
+    metric_variance: variance of the metric
+    mde: absolute minimum detectable effect
+    alpha: statistical significance
+    power: statistical power
+    """
+
+    def __init__(
+        self,
+        metric_variance: float,
+        mde: float,
+        alpha: float = DEFAULT_ALPHA,
+        power: float = DEFAULT_POWER,
+    ):
+        self.metric_variance = metric_variance
+        self.mde = mde
+        self.alpha = alpha
+        self.power = power
+
+
+class SampleSizeCalculator:
+    """This class is to calculate sample size based on variable type
+
+    Attributes:
+    alpha: statistical significance
+    power: statistical power
+
+    """
+
+    def __init__(self, alpha: float = DEFAULT_ALPHA, power: float = DEFAULT_POWER):
+        self.alpha = alpha
+        self.power = power
+        self.boolean_metrics: List[PowerAnalysisParameters] = []
+        self.numeric_metrics: List[PowerAnalysisParameters] = []
+        self.ratio_metrics: List[PowerAnalysisParameters] = []
+
+    def register_bool_metric(self, mde: float, probability: float) -> None:
+        variable = Boolean(probability)
+        metric = PowerAnalysisParameters(variable.variance, mde, self.alpha, self.power)
+        self.boolean_metrics.append(metric)
+
+    def register_numeric_metric(self, mde: float, variance: float) -> None:
+        variable = Numeric(variance)
+        metric = PowerAnalysisParameters(variable.variance, mde, self.alpha, self.power)
+        self.numeric_metrics.append(metric)
+
+    def register_ratio_metric(
+        self,
+        mde: float,
+        numerator_mean: float,
+        numerator_variance: float,
+        denominator_mean: float,
+        denominator_variance: float,
+        covariance: float,
+    ) -> None:
+        variable = Ratio(numerator_mean, numerator_variance, denominator_mean, denominator_variance, covariance)
+        metric = PowerAnalysisParameters(variable.variance, mde, self.alpha, self.power)
+        self.ratio_metrics.append(metric)
+
+    @staticmethod
+    def _get_single_sample_size(metric: PowerAnalysisParameters, power_analysis_type: Any) -> float:
+        effect_size = metric.mde / float(np.sqrt(metric.metric_variance))
+        power_analysis = power_analysis_type()
+        sample_size = int(
+            power_analysis.solve_power(
+                effect_size=effect_size,
+                alpha=metric.alpha,
+                power=metric.power,
+                ratio=1,
+                alternative="two-sided",
+            )
+        )
+        return sample_size
+
+    def get_overall_sample_size(self) -> float:
+        # Supports the sample size calculation for single metric now.
+        # The current structure is set up to support multiple metrics in the future.
+        sample_size = float("nan")
+        if self.boolean_metrics:
+            for metric in self.boolean_metrics:
+                sample_size = self._get_single_sample_size(metric, NormalIndPower)
```

Comment:
Will a particular metric ever have multiple `power_analysis_types`? Wonder if this could be represented in the `Boolean`, `Ratio`, and `Proportion` classes? If that was the case, this entire method could be represented as 

```python3
 def get_overall_sample_size(self) -> float:
        return max(map(self._get_single_sample_size, self.metrics))
```

because we would no longer need to differentiate `self.numeric_metrics`, `self.boolean_metrics`, and `self.ratio_metrics`

Reply:
Good question. For now, we support single metric per calculator, I don't think we will ever introduce multiple power analysis type per variable. We may need multiple correction methods when we support multiple metrics per calculator. I can put the default single metric power analysis type in variable class to make this look more obvious.